libraries/AP_Filesystem/AP_Filesystem_FlashMemory_LittleFS.cpp

```diff
@@ -421,10 +421,37 @@ uint32_t AP_Filesystem_FlashMemory_LittleFS::bytes_until_fsync(int fd)
         return 0;
     }
 
-    uint32_t write_amt = fs_cfg.block_size;
+    uint32_t file_pos = fp->file.pos;
+    uint32_t block_size = fs_cfg.block_size;
+
     // calculate how much allowed if a full block would be written
-    sync_block(fd, fp->file.pos, write_amt);
-    return write_amt; // return that amount
+    uint32_t nbytes = fs_cfg.block_size;
+
+    // see https://github.com/littlefs-project/littlefs/issues/564#issuecomment-2363032827
+    // n = (N − w/8 ( popcount( N/(B − 2w/8) − 1) + 2))/(B − 2w/8))
+    // off = N − ( B − 2w/8 ) n − w/8popcount( n )
+#define BLOCK_INDEX(N, B) \
+    (N - sizeof(uint32_t) * (__builtin_popcount(N/(B - 2 * sizeof(uint32_t)) -1) + 2))/(B - 2 * sizeof(uint32_t))
+
+#define BLOCK_OFFSET(N, B, n) \
+    (N - (B - 2*sizeof(uint32_t)) * n - sizeof(uint32_t) * __builtin_popcount(n))
+
+    uint32_t block_index = BLOCK_INDEX(file_pos, block_size);
+    uint32_t block_offset = BLOCK_OFFSET(file_pos, block_size, block_index);
+    if (block_size - block_offset <= nbytes) {
+        if (block_size == block_offset) {
```

Comment:
So in this path block_size is returned rather than false for no sync, I don't see how the previous PR can have been correct if this is correct.